### ISSUE TITLE
Better support for gradual migrations: better tooling for `CREATE INDEX CONCURRENTLY`

### ISSUE THREAD

**Issue by mzabani** [mzabani/codd#181]
Say one wants to add a new column to a table and populate it gradually, and they want to index it but not pay the price of locking the table due to `CREATE INDEX` in their next deployment. Because of that, they want to use `CREATE INDEX CONCURRENTLY`.

But they don't want to put either of those statements (_CONCURRENTLY_ or otherwise) in a codd migration because the table is large so it could take long, keeping the application down longer than necessary during the deployment.

How would one get the best of all worlds in this case? Currently, it would have to be something along these lines:

1 - One PR with `ALTER TABLE .. ADD COLUMN ..` and extra code to gradually populate the new column.
2 - Someone runs `CREATE INDEX CONCURRENTLY` **manually** to the environment where this is deployed, but only **after** deployment of the PR above.
3 - Another PR with `CREATE INDEX CONCURRENTLY` added as a dev-only migration gets merged immediately after step 2 is done. This will make schemas match before future deployments.

There are multiple issues with the strategy above. The main one is that this won't work if there are multiple environments with different deployment schedules, like Staging and Production. It would work for Staging, but Production would face a schema mismatch because the index will be in the disk representations but not in the pending migrations.

Even for a single environment, the need for lockstep actions and manual application of SQL is unacceptable.

We need to think of tooling to avoid manual interventions and still get all the positives above in a single PR.

Todo: WIP

**Comment by mzabani** [comment]
We can maybe simplify this further and ask for migrations in pairs. Following our example of adding a column and indexing it concurrently, the migration that adds the column is just a normal migration anyway, so we don't need to care for it in our analysis.

So special migrations would come in pairs, e.g.:

```sql
-- codd: begin-concurrent-migration
-- codd: no-txn

Create index concurrently on table(col);
```

Codd would include the index in on-disk representations to better represent schema state after migrations are all applied, but would mark any files affected by the migration above with its name and timestamp so it can know to ignore a file or not when verifying representations. Let us worry only about concurrent migrations that create rep files, not ones that modify or delete them.
Codd should also include whether indexes are valid in its representations (if it doesn't yet) because that will help ensure deployments create valid indexes or generate an actionable error. Without ensuring no errors happened in all deployed environments, the developer cannot proceed with other tasks and must try to recreate the index, of course. Also, the developer must wait because for large databases these operations can take a while, so waiting until every environment received the update is unavoidable.

Once this is deployed to all environments without errors, the developer would then merge their second PR:
```sql
-- codd: end-concurrent-migration yyyy-mm-dd-....sql 

-- Would this be an empty migration? We might have to allow it
```

This would change the marked rep files back to "normal" on `codd add`.

This would allow for "regular" migrations to e.g. rename the index or drop it to come afterwards. Summarizing `codd up`:
- If only the first migration exists and is pending, apply all "regular" pending migrations first and commit, then apply the `begin-concurrent-migration` last regardless of timestamps.
    - `codd up` is still idempotent
    - Other regular migrations can be added as long as they don't delete or modify rep files marked by concurrent migrations (a migration renaming the index would fail even before verifying checksums because it would run before the special migration creates the index)
- If both migrations are pending, codd will just apply them as if they were regular migrations and obey timestamp order.
   - `codd up` is still idempotent.
   - Other regular migrations can now be added and manipulate the index freely.
- If only the second migration is pending, apply it like a regular migration.

What if index creation fails in some environment? Then the developer would have to merge a new PR with two migrations to try again:

```sql
-- codd: end-concurrent-migration yyyy-mm-dd-....sql 

Drop index ...;
```

And once again:
```sql
-- codd: begin-concurrent-migration
-- codd: no-txn

Create index concurrently on table(col);
```

A similar approach using `REINDEX CONCURRENTLY ` should be possible, too.

In the case above, the end-concurrent-migration would run like a regular migration and the other one would run last, and that should be fine.

Of course, good state machine tests that emulate merging of PRs from multiple devs and deployment to multiple environments are a must for a feature like this.
